sdk/messaging/azservicebus/amqp_message.go

```diff
@@ -0,0 +1,279 @@
+	Durable bool
+
+	// FirstAcquirer corresponds to the 'first-acquirer' property.
+	FirstAcquirer bool
+
+	// Priority corresponds to the 'priority' property.
+	Priority uint8
+
+	// TTL corresponds to the 'ttl' property.
+	TTL time.Duration
+}
+
+// toAMQPMessage converts between our (azservicebus) AMQP message
+// to the underlying message used by go-amqp.
+func (am *AMQPMessage) toAMQPMessage() *amqp.Message {
+	var header *amqp.MessageHeader
+
+	if am.Header != nil {
+		header = &amqp.MessageHeader{
+			DeliveryCount: am.Header.DeliveryCount,
+			Durable:       am.Header.Durable,
+			FirstAcquirer: am.Header.FirstAcquirer,
+			Priority:      am.Header.Priority,
+			TTL:           am.Header.TTL,
+		}
+	}
+
+	var properties *amqp.MessageProperties
+
+	if am.Properties != nil {
+		properties = &amqp.MessageProperties{
+			AbsoluteExpiryTime: am.Properties.AbsoluteExpiryTime,
+			ContentEncoding:    am.Properties.ContentEncoding,
+			ContentType:        am.Properties.ContentType,
+			CorrelationID:      am.Properties.CorrelationID,
+			CreationTime:       am.Properties.CreationTime,
+			GroupID:            am.Properties.GroupID,
+			GroupSequence:      am.Properties.GroupSequence,
+			MessageID:          am.Properties.MessageID,
+			ReplyTo:            am.Properties.ReplyTo,
+			ReplyToGroupID:     am.Properties.ReplyToGroupID,
+			Subject:            am.Properties.Subject,
+			To:                 am.Properties.To,
+			UserID:             am.Properties.UserID,
+		}
+	} else {
+		properties = &amqp.MessageProperties{}
+	}
+
+	var footer amqp.Annotations
+
+	if am.Footer != nil {
+		footer = (amqp.Annotations)(am.Footer)
+	}
+
+	return &amqp.Message{
+		Annotations:           copyAnnotations(am.MessageAnnotations),
+		ApplicationProperties: am.ApplicationProperties,
+		Data:                  am.Body.Data,
+		DeliveryAnnotations:   amqp.Annotations(am.DeliveryAnnotations),
+		DeliveryTag:           am.DeliveryTag,
+		Footer:                footer,
+		Header:                header,
+		Properties:            properties,
+		Sequence:              am.Body.Sequence,
+		Value:                 am.Body.Value,
+	}
+}
+
+func copyAnnotations(src map[any]any) amqp.Annotations {
+	if src == nil {
+		return amqp.Annotations{}
+	}
+
+	dest := amqp.Annotations{}
+
+	for k, v := range src {
+		dest[k] = v
+	}
+
+	return dest
+}
+
+func newAMQPMessage(goAMQPMessage *amqp.Message) *AMQPMessage {
+	var header *AMQPMessageHeader
+
+	if goAMQPMessage.Header != nil {
+		header = &AMQPMessageHeader{
+			DeliveryCount: goAMQPMessage.Header.DeliveryCount,
+			Durable:       goAMQPMessage.Header.Durable,
+			FirstAcquirer: goAMQPMessage.Header.FirstAcquirer,
+			Priority:      goAMQPMessage.Header.Priority,
+			TTL:           goAMQPMessage.Header.TTL,
+		}
+	}
+
+	var properties *AMQPMessageProperties
+
+	if goAMQPMessage.Properties != nil {
+		properties = &AMQPMessageProperties{
+			AbsoluteExpiryTime: goAMQPMessage.Properties.AbsoluteExpiryTime,
```

Comment:
If `amqp.MessageProperties` and `AMQPMessageProperties` have the exact same layout you can cast one to the other and let the compiler do the copying for you.

Reply:
We'll need to do that separately. They have the same content, but their field layout is in a different order which appears to be enough to tank the auto-copying.

I don't think we did that on purpose in go-amqp, so we can patch it there and then vendor that change back into here.